Added logging infrastructure. Fixes #158.

RobertJacobsonCDC opened the pull request:
Added `trace!` statements to `basic-infection` and enabled logging.
Added `trace!` statements throughout Ixa proper.
Added a default set of module log level filters, as it turns out rustyline logs trace messages, which are noisy. Modified `examples/basic-infection` to accept command line arguments.

RobertJacobsonCDC committed:
Added `trace!` statements to `basic-infection` and enabled logging.
Added `trace!` statements throughout Ixa proper.
Added a default set of module log level filters, as it turns out rustyline logs trace messages, which are noisy.
Modified `examples/basic-infection` to accept command line arguments.

RobertJacobsonCDC commented:
I have a draft for log-related debugger commands on the `RobertJacobsonCDC_158_logging_debugger` branch.

RobertJacobsonCDC commented:
**Some notes about this implementation.**

A natural question is whether logging configuration should be stored in a `DataPlugin` and attached to a `Context` instance. A logger is global, not specific to a `Context` instance, and there doesn't seem to be an advantage to accessing logging facilities through a `Context`.  Consequently, logging configuration isn't stored in a `Context`.

Another question is how to balance flexibility and features against simplicity. This question recommends having an API that is agnostic to backend implmentation. There is a complexity cliff in the logging ecosystem: There are the simple "just works" crates that require just a few lines, and there are these byzantine crates with loads of abstractions and features. If we need more features, we can swap out the backend without having to refactor client code.

RobertJacobsonCDC commented on src/log.rs:
```diff
@@ -0,0 +1,214 @@
+// Default module specific filters
+const DEFAULT_MODULE_FILTERS: [(&str, LevelFilter); 1] = [
+    ("rustyline", LevelFilter::Off),
+    // ("ixa", LevelFilter::Off),
```
Since `rustyline` logs noisy `trace` messages, I think we need to have `rustyline` specifically filtered out by default. But it's not clear if we should have internal `ixa` messages filtered out by default if the user enables logging. I would think not, but...

RobertJacobsonCDC commented on src/log.rs:
```diff
@@ -0,0 +1,214 @@
+
+/// Sets the level filters for a set of modules according to the provided map. Use this instead of
+/// `set_module_filter()` to set filters in bulk.
+#[allow(clippy::implicit_hasher)]
```
Clippy has a lint that wants every `HashMap` to be generic on the `HashBuilder`. (?! That seems bonkers to me. We wouldn't require everything that allocates to be generic on the allocator, would we?) There is no circumstance I can imagine in which client code would want a specialty hasher for this use case.

k88hudson-cfa commented on src/log.rs:
```diff
@@ -0,0 +1,214 @@
+
+/// Enables the logger with no global level filter / full logging. Equivalent to
+/// `set_log_level(LevelFilter::Trace)`.
+pub fn enable_logging() {
```
I'll add a comment in the runner section but I think I'd probably be in favor of keeping the programmatic API simple (like just `set_log_level`, not `enable_logging`/`disable_logging`) and having the syntactic sugar be at the command line

RobertJacobsonCDC replied:
Do you mean have `set_module_filter` and others as a command line argument? Or do you mean leave the command line argument as it is (just `--enable-logging`) and leave out all of the programmatic API except for `set_log_level`?

RobertJacobsonCDC replied:
(Clarified in chat.)

k88hudson-cfa commented on src/log.rs:
```diff
@@ -0,0 +1,214 @@
+            }
+        }
+    }
+}
```
Could you maybe add a couple of tests?

RobertJacobsonCDC replied:
I knew I was forgetting something. 🤦

k88hudson-cfa commented on src/runner.rs:
```diff
@@ -43,6 +49,7 @@ impl BaseArgs {
             output_dir: None,
             file_prefix: None,
             force_overwrite: false,
+            enable_logging: false,
```
This definitely a matter of taste but I think I would lean towards having `error` on by default, and then `--log-level={level}`. I could also see cargo-style  `-v` or `--verbose` (maybe for debug?) / `-vv` or `--very-verbose` (for trace) 

As a follow-up: after looking at your comments with `rustyline` etc. above, I do think we are going to want to allow module-level log settings if people need them (either through a command line option or and environment variable)

k88hudson-cfa reviewed (changes requested):
I think as a general principle conventional is good here in terms of the interface. This will be good to put in front of people and see what they think :)

RobertJacobsonCDC committed:
Added tests and rephrased some doc comments.
Modified the command line argument for logging to set the log level filter rather than just be an on-off switch.

RobertJacobsonCDC committed:
Changed from `env_logger` to `log4rs`.

RobertJacobsonCDC commented:
@k88hudson-cfa Rebased and switched from `env_logger` to `log4rs`, which supports changing level filters at runtime.

k88hudson-cfa commented on src/log.rs:
```diff
@@ -0,0 +1,360 @@
+fn get_log_configuration() -> MutexGuard<'static, LogConfiguration> {
+    // Silence lint about mutable global variables.
+    #[allow(static_mut_refs)]
+    unsafe {
```
Can you actually use a `RefCell` here instead (like something similar to what we're doing with `GLOBAL_PROPERTIES`)? I was hoping we could avoid adding `unsafe` here

RobertJacobsonCDC replied:
🤦 That's vestigial. I've removed all unsafe blocks. The globals don't need to be mut.

k88hudson-cfa reviewed (changes requested):
API-wise I think this looks good, thanks for reworking it – I had a comment about mutable globals, othweise I think this is good to land

k88hudson-cfa commented on src/log.rs:
```diff
@@ -0,0 +1,360 @@
+    use log::{error, trace, LevelFilter};
+    use std::sync::{LazyLock, Mutex};
+
+    // Force logging tests to run serially for consistent behavior.
```
Would something like this be a good idea maybe? You'd probably know better than me on what's normal here https://crates.io/crates/serial_test

RobertJacobsonCDC replied:
I think the simplest and easiest thing is the best thing. The `serial_test` crates adds another dependency and just uses a mutex under the hood anyway.

But my mutex doesn't need to be mut, so I don't need unsafe.

k88hudson-cfa reviewed (approved):
🚢 it
